proptest/src/test_runner/errors.rs

```diff
@@ -33,9 +33,32 @@ pub enum TestCaseError {
     Fail(Reason),
 }
 
+/// Indicates the type of test that ran successfully.
+///
+/// This is used for managing whether or not a success is counted against
+/// configured `PROPTEST_CASES`; only `NewCases` shall be counted.
+///
+/// TODO-v2: Ideally `TestCaseResult = Result<TestCaseOk, TestCaseError>`
+/// however this breaks source compability in version 1.x.x because
+/// `TestCaseResult` is public.
+#[derive(Debug, Clone)]
+pub(crate) enum TestCaseOk {
+    NewCaseSuccess,
+    ReplaySuccess,
+    CacheHitSuccess,
+    Reject,
+}
```

Comment:
Not expecting any changes inline to this PR as these thoughts aren't fully baked but wanted to run some stuff by you and see how you're thinking of these things. Also since this is a crate-private type, we can evolve this independently of crate version so that gives us some flexibility for the future.

I've been thinking of how to add some new functionality to support #284 (providing manual, explicit cases) and was thinking of introducing a new wrapper type for cases that carry context on their source so we can have something like:
```
enum Case<T: fmt::Debug, S: Strategy<Value=T>> {
    UserProvided(T),    
    Generated(S),
    Persisted(S, Seed),
    Replay(S, TestCaseResult),
}
```

or something along these lines and then the test-runner code can branch on these accordingly to affect seed, success counts, whether we actually run the test or just emulate it based on replay value, etc.

We could be left with a result type of:
```
enum TestCaseOk {
    Success,
    CacheHit,
    Reject,
}
```

Not entirely sure all this would end up working out with the current code structure as I haven't actually started implementing it.

Reply:
And more direct comment on this code after reading some more below, I think we'd want a `PersistedCaseSuccess` variant. Right now it seems `ReplaySuccess` is being overloaded for both persisted tests as well as forked tests.

Reply:
that seems like a good approach to me 👍